benchmarks/dataset_serialize_benchmark.py

```diff
@@ -0,0 +1,232 @@
+    a file system backed by RAM (tmpfs). The assumption is that writing to
+    tmpfs is fast (so fast that benchmark duration is significantly affected by
+    the serialization itself), and stable (so that its performance is ~constant
+    across runs on the same machine).
+
+    This benchmark does not resolve how much time goes into the CPU work for
+    serialization vs. the system calls for writing to tmpfs (that would be a
+    different question to answer, an interesting one, that is maybe more of a
+    task for profiling).
+
+    There are two dimensions that are varied:
+
+        - serialization format
+        - amount of the data being written, as set by a filter on the input
+    """
+
+    name = "dataset-serialize"
+
+    arguments = ["source"]
+
+    sources = [
+        "nyctaxi_multi_parquet_s3",
+        # "nyctaxi_multi_ipc_s3",
+        # "chi_traffic_2020_Q1",
+    ]
+
+    sources_test = [
+        "nyctaxi_multi_parquet_s3_sample",
+        "nyctaxi_multi_ipc_s3_sample",
+        "chi_traffic_sample",
+    ]
+
+    _params = {
+        "selectivity": ("1pc", "10pc"),
+        "format": (
+            "parquet",
+            "arrow",
+            "ipc",
+            "feather",
+            "csv",
+        ),
+    }
+
+    valid_cases = [tuple(_params.keys())] + list(
+        itertools.product(*[v for v in _params.values()])
+    )
+
+    filters = {
+        "nyctaxi_multi_parquet_s3": {
+            "1pc": ds.field("pickup_longitude") < -74.013451,  # 561384
+            "10pc": ds.field("pickup_longitude") < -74.002055,  # 5615432
+            "100pc": None,  # 56154689
+        },
+        "nyctaxi_multi_ipc_s3": {
+            "1pc": ds.field("pickup_longitude") < -74.014053,  # 596165
+            "10pc": ds.field("pickup_longitude") < -74.002708,  # 5962204
+            "100pc": None,  # 59616487
+        },
+        "chi_traffic_2020_Q1": {
+            "1pc": ds.field("END_LONGITUDE") < -87.807262,  # 124530
+            "10pc": ds.field("END_LONGITUDE") < -87.7624,  # 1307565
+            "100pc": None,  # 13038291
+        },
+        **dict.fromkeys(
+            ["nyctaxi_multi_parquet_s3_sample", "nyctaxi_multi_ipc_s3_sample"],
+            {
+                "1pc": ds.field("pickup_longitude") < -74.0124,  # 20
+                "10pc": ds.field("pickup_longitude") < -74.00172,  # 200
+                "100pc": None,  # 2000
+            },
+        ),
+        "chi_traffic_sample": {
+            "1pc": ds.field("END_LONGITUDE") < -87.80726,  # 10
+            "10pc": ds.field("END_LONGITUDE") < -87.76148,  # 100
+            "100pc": None,  # 1000
+        },
+    }
+
+    _case_tmpdir_mapping = {}
+
+    def _create_tmpdir_in_ramdisk(self, case: tuple):
+        # Build simple prefix string to facilitate correlating directory names
+        # to test cases.
+        pfx = "-".join(c.lower()[:9] for c in case)
+        dirpath = os.path.join("/dev/shm", pfx + "-" + str(uuid.uuid4()))
```

Comment:
Can we rely on /dev/shm existing?

Reply:
On Linux we can, yes.

Maybe this is interesting background: https://unix.stackexchange.com/a/151987/13256

On MacOS, sadly not:

> Support for POSIX shared memory is somewhat limited. Though the functionality is implemented, there is no file system interface like /dev/shm to manage the shared memory segments.

Maybe we should mark this test to be skipped on MacOS.

Reply:
It's not a test, exactly. The other situation like this I know we have is file cache clearing; we try a command that only works on linux, but don't fail if it fails (as a lot of us develop on macos). It'd be nice to at least be able to run this locally for dev purposes if possible, even if this is how we want it to run in CI.